src/cpu/aarch64/acl_inner_product.cpp

```diff
@@ -0,0 +1,67 @@
+/*******************************************************************************
+* Copyright 2021 Arm Ltd. and affiliates
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "cpu/aarch64/acl_inner_product.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace cpu {
+namespace aarch64 {
+
+using namespace dnnl::impl::status;
+using namespace dnnl::impl::memory_tracking::names;
+using namespace dnnl::impl::utils;
+
+status_t acl_inner_product_fwd_t::execute_forward(const exec_ctx_t &ctx) const {
+    status_t status = status::success;
+    auto src_base = CTX_IN_MEM(const data_t *, DNNL_ARG_SRC);
+    auto wei_base = CTX_IN_MEM(const data_t *, DNNL_ARG_WEIGHTS);
+    auto bia_base = CTX_IN_MEM(const data_t *, DNNL_ARG_BIAS);
+    auto dst_base = CTX_OUT_MEM(data_t *, DNNL_ARG_DST);
+
+    bool with_bias = pd()->aip_.with_bias;
+    bool with_sum = pd()->aip_.with_sum;
+
+    // Retrieve primitive resource and configured Compute Library objects
+    auto *acl_resource
```

Comment:
As you mentioned there is a potential concurrency problem. As I understand `resource_mapper_t` is a field of a `primitive_t`. If multiple threads execute the same primitive object then both threads will access the same resource and will try to modify it. So there no issues between primitives, but there is an issue within a single primitive execution in multithread environment. This affects whole library because of `resource_mapper_t` design. Summoning @densamoilov in case I am missing something.

Reply:
>but there is an issue within a single primitive execution in multithread environment

There is a case of `convolution_fwd` primitive, which works fine in multithreaded environment. Or is Conv2D specical?

Reply:
Convolution has the same issue, I am sure we don't have tests where a primitive is executed by multiple threads concurrently. 

Resources were designed as objects which are constant during execution, but which can't be placed into primitive cache, because they are not constant during execution or we can't represent them as part of key in primitive cache. For example, on GPU primitives use resources for preparing and storing memory objects with zero_points, scale, etc. prior to execution. On CPU resources are not used anywhere except ARM implementation. This is probably fine until we have a request to support concurrent execution of a primitive on ARM.

Reply:
Makes sense. As I understand you could also add a guard mutex in `execute()` function, so even 2 threads call `primitive::execute()` their execution will be sequential.

Reply:
So far I have not encountered problems with this primitive, but this is a good point, thanks.

Reply:
In TensorFlow each (inter-)pthread has its own set of primitives and the sets of different threads don't intersect, thus there are no concurrency issues with `resource_mapper_t` in the case of TF. However, somebody can always design a custom implementation of multithreaded primitive execution, thus to be safe it seems better to add a lock guard in the scope of `execute()` which I plan to do in a separate PR.

Reply:
Sounds good to me, thanks for investigating the issue!